README.md
# cointracking_polka

## 1 Overview

Staking in Polkadot is a tax-relevant process. But how do I find out what income
I have generated?  
Maybe you already use <https://cointracking.info>, the
"Leader for Cryptocurrency Tracking and Tax Reporting".  
If you do so, you have to import your staking data in the Cointracking Service.

This is where **cointracking_polka** comes into play, it generates reports on
staking income and fees for accounts to be freely defined.

**cointracking_polka** is a Node.js program written in typescript to create
staking reports usable with <https://cointracking.info>, the service which helps
you with tax return.  
The created csv files can be imported in the cointracking database using the menu:  
*Enter Coins | Bulk Imports | CSV Import*

The following coins are supported:

* Polkadot
* Kusama
* Westend (testnet, therefore not relevant for cointracking)

## 2 Installation

### 2.1 Prerequisites

These steps should only be carried out during the initial installation.

#### 2.1.1 Repository

First you have to clone the repository:

``` bash
git clone https://github.com/TheGoldenEye/cointracking_polka.git
```

#### 2.1.2 Needed packages

We need some prerequisites:

``` bash
sudo apt install node-typescript npm
```

#### 2.1.3 Minimum node.js version

Now its time to check the nodejs version:

``` bash
node -v
```

If your node version is minimum v10.4.0, its fine. Otherwise you have to install
a newer version, because of the missing BigInt support in Node.js prior to v10.4.  
You can do it with the 'n node installer':

``` bash
sudo npm install -g n
sudo n lts
```

Now you should have a current node version installed.

#### 2.1.4 yarn package manager

This repo uses yarn workspaces to organise the code.
As such, after cloning, its dependencies should be installed via yarn package manager,
not via npm, the latter will result in broken dependencies.  
Install yarn, if you haven't already:

``` bash
sudo npm install -g yarn
```

### 2.2 Installing the project dependencies

*Please always run this when the sources have been updated from the git repository.*

Use yarn to install the dependencies:

``` bash
cd cointracking_polka
yarn
```

### 2.3 Installing the polka-store database(s)

The tool requires access to all transactions of the accounts to be analyzed.
Unfortunately, this is not possible in the Polkadot universe, since transaction
data cannot be queried directly there.  
Therefore **cointracking_polka** uses the transaction database of **polka-store**
<https://github.com/TheGoldenEye/polka-store>. Polka-store scans a Polkadot chain
(Polkadot/Kusama/Westend) and stores balance-relevant transactions in a SQLite database.  
If you want, you can download the latest versions of the transaction databases
[here](https://e.pcloud.link/publink/show?code=kZx3eZENGTspnf6YLueJK6F2w8ULTpnFIk).  
Alternatively you can install **polka-store** and create the transaction databases yourself.  

*Please copy the Polkadot.db and/or Kusama.db to the data directory.*

## 3 Configuration

### 3.1 config/config.json

You can find a template for configuration in `config/config_tpl.json`.
Instead, the file used by the tool is `config/config.json`.
If the file does not exist, `config/config_tpl.json` is copied to
`config/config.json` during the program start phase.  
You can adapt this file (`config.json`) to your needs.

Here are the parameters defined for the different chains.  
In the `accounts` list you can define the accounts to analyze (stash accounts
for staking rewards or any other accounts for tracking the fees).  
Besides the accounts list there is currently no need to change the default configuration.

``` json
config.json:
{
  "staking": "day",
  "feeReceived": "day",
  "feePaid": "day",
  "defchain": "Polkadot",
  "chains": {
    "Polkadot": {
      "database": "data/Polkadot.db",
      "unit": "DOT",
      "ticker": "DOT2",
      "decimals": 10,
      "accounts": [
        { "name": "Example1", "account": "15kUt2i86LHRWCkE3D9Bg1HZAoc2smhn1fwPzDERTb1BXAkX" },
        { "name": "Example2", "account": "12xtAYsRUrmbniiWQqJtECiBQrMn8AypQcXhnQAc6RB6XkLW" }
      ]
    },
    "Kusama": {
      "database": "data/Kusama.db",
      "unit": "KSM",
      "ticker": "KSM",
      "decimals": 12,
      "accounts": [
        { "name": "Example1", "account": "HmFYPT1btmi1T9qqs5WtuNJK93yNdnjjhReZh6emgNQvCHa" },
        { "name": "Example2", "account": "GXPPBuUaZYYYvsEquX55AQ1MRvgZ96kniEKyAVDSdv1SX96" }
      ]
    },
    "Westend": {
      "database": "data/Westend.db",
      "unit": "WND",
      "ticker": "WND",
      "decimals": 12,
      "accounts": [
        { "name": "Example1", "account": "5FnD6fKjTFLDKwBvrieQ6ZthZbgMjppynLhKqiRUft9yr8Nf" },
        { "name": "Example2", "account": "5DfdW2r2hyXzGdXFqAVJKGrtxV2UaacnVNr3sAdgCUDc9N9g" }
      ]
    }
  },
  "csv": {
    "header":      "\"Type\",\"Buy\",\"Cur.\",\"Sell\",\"Cur.\",\"Fee\",\"Cur.\",\"Exchange\",\"Group\",\"Comment\",\"Date\",\"TradeID\",\"BuyValue\",\"SellValue\"",
    "staking":     "\n\"Staking\",\"%s\",\"%s\",,,,,\"Staking_%s\",\"%s\",\"stash: %s tx:%s\",\"%s\",\"Staking_%s_%s\"",
    "feeReceived": "\n\"Staking\",\"%s\",\"%s\",,,,,\"Staking_%s\",\"%s\",\"Fee received: tx:%s\",\"%s\",\"Staking_%s_%s\"",
    "feePaid":     "\n\"Trade\",\"0\",\"EUR\",\"%s\",\"%s\",\"%s\",\"%s\",\"Staking_%s\",\"%s\",\"Fee paid: tx:%s\",\"%s\",\"Staking_%s_%s\",\"0.00000001\",\"0.00000001\""
  }
}
```

**_Global settings:_**  
**staking:**  Please define the aggregation level of the staking records.  
Valid values: [ "individual", "time", "day", "none" ]

* "individual": no aggregation
* "time":  all staking records with the same time are combined
* "day":  all staking records of a day are combined
* "none": disables staking records

**feeReceived:**  Please define the aggregation level of the records based on
fees you got as block author (validators only).  
Valid values: [ "individual", "time", "day", "none" ]  
The values have the same meaning as in staking  
**feePaid:**  Please define the aggregation level of the records based on
fees you paid in various transactions.  
Valid values: [ "individual", "time", "day", "none" ]  
The values have the same meaning as in staking  
**defchain:** The chain which is used (if no chain is given in the command line)  

**_Chain specific settings:_**  
**database:** The path to the sqlite transaction database (input) from polka-store  
**unit:** The unit of the coin  
**ticker:** The associated ticker symbol in cointracking  
**decimals:** Defines the number of decimal places  
**accounts:** A list of accounts to analyze

## 4 Running

### 4.1 Compile Typescript

Now you have to build the code (compile typescript to javascript)

``` bash
yarn build
```

### 4.2 Start program

**One** of the following commands starts the tool, collecting data from the **given** chain:

``` bash
yarn polkadot
yarn kusama
```

Your console will show information like this:

``` text
---------------------------------------------------------------
cointracking-polka: v1.0.0
Chain:              Polkadot
---------------------------------------------------------------
Using transaction data until block: 3139419 (2020-12-31 15:39:06)
Query data from database...
  Progress: 100%
Write csv file...
  Progress: 100%

Staking records:       112   Total: 94.5173137605 DOT
FeeReceived records:     0   Total: 0.0000000000 DOT
FeePaid records:        27   Total: -0.7801000000 DOT
```

## 5 CSV Output

The created csv file you can find in the *output* directory. You con import it
in <https://cointracking.info> using the menu:  
*Enter Coins | Bulk Imports | CSV Import*

## 6 Known issues

* Currently no known issues

## 6 Contributions

I welcome contributions. Before submitting your PR, make sure to run the following commands:

* `yarn lint`: Make sure your code follows the linting rules.
* `yarn lint --fix`: Automatically fix linting errors.

<https://github.com/TheGoldenEye/cointracking_polka/graphs/contributors>

## 7 Authors

* GoldenEye

## 8 Please support me

If you like my work, please consider to support me in the Polkadot/Kusama
networks and nominate my validators:

**Polkadot:**

1. [Validator GoldenEye](https://polkadot.subscan.io/account/14K71ECxvekU8BXGJmSQLed2XssM3HdBYQBuDUwHeUMUgBHk)
2. [Validator GoldenEye/2](https://polkadot.subscan.io/account/14gYRjn6fn5hu45zEAtXodPDbtaditK8twoWUXFi6DsLwd31)

**Kusama:**

1. [Validator GoldenEye](https://kusama.subscan.io/account/FiNuPk2iPirbKC7Spse3NuE9rWjzaQonZmk6wRvk1LcEU13)
2. [Validator GoldenEye/2](https://kusama.subscan.io/account/GcQXL1HgF1ZETZi3Tw3PoXGWeXbDpfsJrrgNgwxde4uoVaB)
3. [Validator GoldenEye/3](https://kusama.subscan.io/account/HjH4dvyPv2RQMA6XUQPqF37rZZ8seNjPQqYRSm3utdszsin)

## 9 License

Apache 2.0 License  
Copyright (c) 2021 GoldenEye

**Disclaimer:
The tool is provided as-is. I cannot give a guarantee for accuracy and I assume NO LIABILITY.**


CHANGELOG.md
# CHANGELOG

## 1.3.0 Mar 9, 2023

Changes:

- Possibility to define account lists for staking and not-staking

## 1.2.0 Dec 29, 2021

Changes:

- Added optional *"Comment"* field in account definition (config.json)
- database queries optimized
- added "createTransfers" flag to create Deposit/Withdrawal records
- allow to define a list of transaction ids to ignore (*"ignoreTx"*) 
- allow to define a list of deposit/withdrawal ids to ignore (*"ignoreTxDW"*) 
- consider renamed event names (e.g. *'staking.Reward'* -> *'staking.Rewarded'*)
- introduce a list of (external) account names, used in deposit/withdrawal records

## 1.1.0 Jan 3, 2021

Changes:

- only use data of complete days (data record aggregation by *'day'*)

## 1.0.0 Jan 2, 2021

Changes:

- Initial release


src/main.ts
// Required imports
import { DivideS, LoadConfigFile, TimeStr, EndOfPeriod } from './utils';
import { sprintf } from "sprintf-js";
import * as fs from 'fs';
import * as getPackageVersion from '@jsbits/get-package-version';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const db = require('better-sqlite3-helper');

// --------------------------------------------------------------
type TStakingRecord = {
  name: string,
  id: string,
  recipientId: string,
  stash: string,
  amount: bigint,
  date: string,
  ignore?: boolean;
}

// --------------------------------------------------------------
type TFeeReceivedRecord = {
  name: string,
  id: string,
  authorId: string,
  feeReceived: bigint,
  date: string,
  ignore?: boolean;
}

// --------------------------------------------------------------
type TFeePaidRecord = {
  name: string,
  id: string,
  senderId: string,
  feePaid: bigint,
  date: string,
  ignore?: boolean;
}

// --------------------------------------------------------------
type TDepositRecord = {
  name: string,
  id: string,
  type: string,
  senderId: string,
  recipientId: string,
  amount: bigint,
  date: string,
  ignore?: boolean;
}

// --------------------------------------------------------------
type TWithdrawalRecord = {
  name: string,
  id: string,
  type: string,
  senderId: string,
  recipientId: string,
  amount: bigint,
  date: string,
  ignore?: boolean;
}

// --------------------------------------------------------------
type TStakefromTo = {
  from: string | null,
  to: string | null
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function CreateCsv(fs: any, header: string, date: string, unit: string): string {
  const fileName = 'output/' + date + '_' + unit + '.csv';

  if (fs.existsSync(fileName))
    fs.unlinkSync(fileName);

  fs.writeFileSync(fileName, header);

  return fileName;
}

// --------------------------------------------------------------
function GetStakingRecords(name: string, accountID: string, startHeight: bigint, atBlock: bigint): TStakingRecord[] {
  const res: TStakingRecord[] = db().query('SELECT ? as name, id, recipientId, addData AS stash, amount, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                            FROM transactions WHERE event=\'staking.Rewarded\' AND addData=? AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);
  return res;
}

// --------------------------------------------------------------
function GetFeePaidRecords(name: string, accountID: string, startHeight: bigint, atBlock: bigint): TFeePaidRecord[] {
  const res1: TFeePaidRecord[] = db().query('SELECT ? as name, id, senderId, totalFee AS feePaid, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                             FROM transactions WHERE totalFee IS NOT NULL AND senderId=? AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);

  // fees for identity jugement:
  const res2: TFeePaidRecord[] = db().query('SELECT ? as name, id, senderId, amount AS feePaid, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                             FROM transactions WHERE event LIKE \'balances.ReserveRepatriated%\' AND senderId=? AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);

  const res = res1.concat(res2);

  // sort res by date
  res.sort((a: TFeePaidRecord, b: TFeePaidRecord) => { return a.date < b.date ? -1 : a.date > b.date ? 1 : 0; });
  return res;
}

// --------------------------------------------------------------
function GetFeeReceivedRecords(name: string, accountID: string, startHeight: bigint, atBlock: bigint): TFeeReceivedRecord[] {
  const res: TFeeReceivedRecord[] = db().query('SELECT ? as name, id, authorId, feeBalances AS feeReceived, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                                FROM transactions WHERE feeBalances IS NOT NULL AND authorId=? AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);
  return res;
}

// --------------------------------------------------------------
function GetDepositRecords(name: string, accountID: string, startHeight: bigint, atBlock: bigint): TDepositRecord[] {
  const res: TDepositRecord[] = db().query('SELECT ? as name, id, type, senderId, recipientId, amount, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                             FROM transactions WHERE senderId IS NOT NULL AND recipientId=? AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);
  return res;
}

// --------------------------------------------------------------
function GetWithdrawalRecords(name: string, accountID: string, startHeight: bigint, atBlock: bigint): TWithdrawalRecord[] {
  const res: TWithdrawalRecord[] = db().query('SELECT ? as name, id, type, senderId, recipientId, amount, datetime(timestamp/1000, \'unixepoch\', \'localtime\') AS date \
                                                FROM transactions WHERE senderId=? AND recipientId IS NOT NULL AND height>=? AND height<=? ORDER BY timestamp ASC',
    name, accountID, startHeight, atBlock);
  return res;
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function GetSpecialStakingAccount(chainData: any, account: string, date: string): string {
  const d = new Date(date);
  let e: TStakefromTo;
  let d1 = new Date(1900, 0);   // 1900-01-01 default, if from is null
  let d2 = new Date(3000, 0);   // 3000-01-01 default, if to is null
  e = chainData.accountsStaking[account];
  if (e) {
    if (e.from)
      d1 = new Date(e.from);
    if (e.to)
      d2 = new Date(e.to);
    return (d1 <= d && d <= d2) ? 'STAKE ' : '';
  }

  e = chainData.accountsNotStaking[account];
  if (e) {
    if (e.from)
      d1 = new Date(e.from);
    if (e.to)
      d2 = new Date(e.to);
    return (d1 <= d && d <= d2) ? 'NOSTAKE ' : '';
  }

  return '';
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function OutputStaking(progressOfs: number, progressCnt: number, file: string, format: string, daily: boolean, chainData: any, arr: TStakingRecord[]): { count: number, value: bigint } {
  const decimals = chainData.decimals;
  const unit = chainData.unit;
  const ticker = chainData.ticker;

  let value = BigInt(0);
  let count = 0;

  arr.forEach((e, idx) => {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', progressOfs + (progressCnt * idx / arr.length)));

    if (e.ignore)
      return;

    value += e.amount;
    count++;

    const val = DivideS(e.amount, decimals);
    const comment_prefix = GetSpecialStakingAccount(chainData, e.recipientId, e.date);

    //"staking":  "\n\"Staking\",\"%s\",\"%s\",,,,,\"Staking_%s\",\"%s\",\"%sstash: %s tx:%s\",\"%s\",\"Staking_%s_%s\"
    fs.appendFileSync(file, sprintf(format, val, ticker, unit, e.recipientId, comment_prefix, e.stash, e.id, daily ? e.date.substring(0, 11) + '23:59:59' : e.date, unit, e.id));
  });

  return {
    count: count,
    value: value
  };
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function OutputFeeReceived(progressOfs: number, progressCnt: number, file: string, format: string, daily: boolean, chainData: any, arr: TFeeReceivedRecord[]): { count: number, value: bigint } {

  const decimals = chainData.decimals;
  const unit = chainData.unit;
  const ticker = chainData.ticker;

  let value = BigInt(0);
  let count = 0;

  arr.forEach((e, idx) => {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', progressOfs + (progressCnt * idx / arr.length)));

    if (e.ignore)
      return;

    value += e.feeReceived;
    count++;

    const val = DivideS(e.feeReceived, decimals);
    const comment_prefix = GetSpecialStakingAccount(chainData, e.authorId, e.date);

    // "\n\"Staking\",\"%s\",\"%s\",,,,,\"Staking_%s\",\"%s\",\"%sFee received: tx:%s\",\"%s\",\"Staking_%s_%s\"",
    fs.appendFileSync(file, sprintf(format, val, ticker, unit, e.authorId, comment_prefix, e.id, daily ? e.date.substring(0, 11) + '23:59:59' : e.date, unit, e.id));
  });

  return {
    count: count,
    value: value
  };
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function OutputFeePaid(progressOfs: number, progressCnt: number, file: string, format: string, daily: boolean, chainData: any, arr: TFeePaidRecord[]): { count: number, value: bigint } {

  const decimals = chainData.decimals;
  const unit = chainData.unit;
  const ticker = chainData.ticker;

  let value = BigInt(0);
  let count = 0;

  arr.forEach((e, idx) => {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', progressOfs + (progressCnt * idx / arr.length)));

    if (e.ignore)
      return;

    value += e.feePaid;
    count++;

    const val = DivideS(e.feePaid, decimals);
    const comment_prefix = GetSpecialStakingAccount(chainData, e.senderId, e.date);

    // "\n\"Trade\",\"0\",\"EUR\",\"%s\",\"%s\",\"%s\",\"%s\",\"Staking_%s\",\"%s\",\"%sFee paid: tx:%s\",\"%s\",\"Staking_%s_%s\",\"0.00000001\",\"0.00000001\""
    fs.appendFileSync(file, sprintf(format, val, ticker, val, ticker, unit, e.senderId, comment_prefix, e.id, daily ? e.date.substring(0, 11) + '23:59:59' : e.date, unit, e.id));
  });

  return {
    count: count,
    value: value
  };
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function OutputDeposits(progressOfs: number, progressCnt: number, file: string, format: string, chainData: any, arr: TDepositRecord[]): number {

  const decimals = chainData.decimals;
  const unit = chainData.unit;
  const ticker = chainData.ticker;

  let count = 0;

  arr.forEach((e, idx) => {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', progressOfs + (progressCnt * idx / arr.length)));

    if (e.ignore)
      return;

    count++;

    const val = DivideS(e.amount, decimals);
    const name = chainData.names[e.senderId];
    const from = name || e.senderId;
    const comment_prefix = GetSpecialStakingAccount(chainData, e.recipientId, e.date);

    if (!name && !chainData.namesNotFound.includes(from)) {
      chainData.namesNotFound.push(from);
    }

    // "\n\"Deposit\",\"%s\",\"%s\",,,,,\"Wallet_%s \",\"%s\",\"%stx:%s from %s\",\"%s\",\"Wallet_%s_%s_D\"",
    fs.appendFileSync(file, sprintf(format, val, ticker, unit, e.recipientId, comment_prefix, e.id, from, e.date, unit, e.id));
  });

  return count;
}

// --------------------------------------------------------------
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function OutputWithdrawals(progressOfs: number, progressCnt: number, file: string, format: string, chainData: any, arr: TWithdrawalRecord[]): number {

  const decimals = chainData.decimals;
  const unit = chainData.unit;
  const ticker = chainData.ticker;

  let count = 0;

  arr.forEach((e, idx) => {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', progressOfs + (progressCnt * idx / arr.length)));

    if (e.ignore)
      return;

    count++;

    const val = DivideS(e.amount, decimals);
    const name = chainData.names[e.recipientId];
    const to = name || e.recipientId;
    const comment_prefix = GetSpecialStakingAccount(chainData, e.senderId, e.date);

    if (!name && !chainData.namesNotFound.includes(to)) {
      chainData.namesNotFound.push(to);
    }

    // "\n\"Withdrawal\",,,\"%s\",\"%s\",,,\"Wallet_%s\" ,\"%s\",\"%stx:%s to %s\",\"%s\",\"Wallet_%s_%s_W\""
    fs.appendFileSync(file, sprintf(format, val, ticker, unit, e.senderId, comment_prefix, e.id, to, e.date, unit, e.id));
  });

  return count;
}

// --------------------------------------------------------------
// --------------------------------------------------------------
function main() {
  // command line parameters:
  // process.argv[2]: chain (optional)
  // process.argv[3]: atBlock (optional)

  const config = LoadConfigFile();
  if (!config)
    return;

  // check given chain
  const chain = process.argv[2] || config.defchain;
  const chainData = config.chains[chain];
  if (!chainData) {
    console.log('Syntax: node build/main.js [chain] [atBlock]');
    const chains = Object.keys(config.chains).join(', ');
    console.log('        with chain in [%s]', chains);
    console.log('        with atBlock a block number');
    return;
  }

  // open database
  const options = {
    path: chainData.database || 'data/' + chain + '.db',
    readonly: false,
    fileMustExist: true, // throw error if database not exists
    WAL: false, // automatically enable 'PRAGMA journal_mode = WAL'?
    migrate: false,
  }
  db(options);
  db().defaultSafeIntegers(true);

  process.on('exit', () => db().close());     // close database on exit

  const ver = getPackageVersion();
  console.log('---------------------------------------------------------------');
  console.log(`cointracking-polka: v${ver}`);
  console.log(`Chain:              ${chain}`);
  console.log('---------------------------------------------------------------');
  if (!chainData.accounts.length) {
    console.log('No accounts given, abort.');
    return;
  }

  //const lastBlock = Number(db().queryFirstRow('SELECT max(height) FROM transactions').height);
  const maxBlock = Number.MAX_SAFE_INTEGER;
  const sBlock = process.argv[3];   // the block number was given by command line?
  const atBlock = isNaN(+sBlock) ? maxBlock : Math.min(+sBlock, maxBlock);
  const lastBlock = Number(db().queryFirstRow('SELECT max(height) As height FROM transactions WHERE height<=?', atBlock).height); // the last block <=atBlock
  let date = db().queryFirstRow('SELECT datetime(timestamp/1000, \'unixepoch\', \'localtime\') as dateStr, timestamp, height \
                                 FROM transactions WHERE height=?', lastBlock);
  if (!date) {
    console.log('Missing Data in Database, abort.');
    return;
  }

  // Create mapping list of accounts with its names
  chainData.names = {};
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  chainData.accountNames.forEach((value: any) => {
    chainData.names[value.account] = value.name;
  });
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  chainData.accounts.forEach((value: any) => {
    chainData.names[value.account] = value.name;
  });

  chainData.namesNotFound = [];

  const flagStaking = config.staking;
  const flagFeePaid = config.feePaid;
  const flagFeeReceived = config.feeReceived;
  const flagCreateTransfers = config.createTransfers || false;
  const arrIgnoreTx: string[] = chainData.ignoreTx;
  const arrIgnoreTxDW: string[] = chainData.ignoreTxDW;
  const formatStaking = config.csv.staking;
  const formatFeeReceived = config.csv.feeReceived;
  const formatFeePaid = config.csv.feePaid;
  const formatDeposit = config.csv.deposit;
  const formatWithdrawal = config.csv.withdrawal;
  const unit = chainData.unit;
  const startHeight: bigint = chainData.startHeight;

  if (flagStaking == 'day' || flagFeePaid == 'day' || flagFeeReceived == 'day') {
    // if aggregation by day, we need the last complete day
    const d1 = new Date(date.dateStr);
    const d2 = EndOfPeriod(new Date(d1.getTime() - 86400000), 'day'); // end of the day before

    const h = db().queryFirstRow('SELECT max(height) as height FROM transactions WHERE timestamp<=?', d2.getTime()).height;

    date = db().queryFirstRow('SELECT datetime(timestamp/1000, \'unixepoch\', \'localtime\') as dateStr, timestamp, height FROM transactions WHERE height=?', h);
  }
  if (!date) {
    console.log('Missing Data in Database, abort.');
    return;
  }

  console.log('Using transaction data until block: %d (%s)', Number(date.height), date.dateStr);

  const file = CreateCsv(fs, config.csv.header, date.dateStr.substring(0, 10), unit);

  let arrStaking: TStakingRecord[] = [];
  let arrFeeReceived: TFeeReceivedRecord[] = [];
  let arrFeePaid: TFeePaidRecord[] = [];
  let arrDeposits: TDepositRecord[] = [];
  let arrWithdrawals: TWithdrawalRecord[] = [];

  // 1. iterate over all accounts and collect all data records
  console.log('Query data from database...');
  for (let i = 0, n = chainData.accounts.length; i < n; i++) {
    process.stdout.write(sprintf('\r  Progress: %d%%  ', 100.0 * i / n));

    const accountID = chainData.accounts[i].account;
    const name = chainData.accounts[i].name;

    if (flagStaking != 'none')
      arrStaking = arrStaking.concat(GetStakingRecords(name, accountID, startHeight, date.height));

    if (flagFeeReceived != 'none')
      arrFeeReceived = arrFeeReceived.concat(GetFeeReceivedRecords(name, accountID, startHeight, date.height));

    if (flagFeePaid != 'none')
      arrFeePaid = arrFeePaid.concat(GetFeePaidRecords(name, accountID, startHeight, date.height));

    if (flagCreateTransfers) {
      arrDeposits = arrDeposits.concat(GetDepositRecords(name, accountID, startHeight, date.height));
      arrWithdrawals = arrWithdrawals.concat(GetWithdrawalRecords(name, accountID, startHeight, date.height));
    }

  }
  process.stdout.write('\r  Progress: 100%\n');

  // 2. remove tx to ignore
  arrStaking.forEach((value: TStakingRecord) => {
    if (arrIgnoreTx.includes(value.id))
      value.ignore = true;
  });
  arrFeeReceived.forEach((value: TFeeReceivedRecord) => {
    if (arrIgnoreTx.includes(value.id))
      value.ignore = true;
  });
  arrFeePaid.forEach((value: TFeePaidRecord) => {
    if (arrIgnoreTx.includes(value.id))
      value.ignore = true;
  });
  arrDeposits.forEach((value: TDepositRecord) => {
    if (arrIgnoreTx.includes(value.id) || arrIgnoreTxDW.includes(value.id))
      value.ignore = true;
  });
  arrWithdrawals.forEach((value: TWithdrawalRecord) => {
    if (arrIgnoreTx.includes(value.id) || arrIgnoreTxDW.includes(value.id))
      value.ignore = true;
  });

  // 3. aggregate entries, if required
  // 3.1 staking entries
  if (flagStaking == 'day' || flagStaking == 'time') {
    arrStaking.forEach((value: TStakingRecord, index: number, array: TStakingRecord[]) => {
      if (value.ignore) // the entry was processed already
        return;

      const d1 = new Date(value.date);
      const d2 = EndOfPeriod(d1, flagStaking);
      let i: number;
      for (i = index + 1; i < array.length && array[i].stash == value.stash; i++) {
        const d = new Date(array[i].date);
        if (d > d2)
          break;
        value.amount += array[i].amount;
        value.id += ' ' + array[i].id;
        array[i].ignore = true;
      }
    });
  }

  // 3.2 feeReceived entries
  if (flagFeeReceived == 'day' || flagFeeReceived == 'time') {
    arrFeeReceived.forEach((value: TFeeReceivedRecord, index: number, array: TFeeReceivedRecord[]) => {
      if (value.ignore) // the entry was processed already
        return;

      const d1 = new Date(value.date);
      const d2 = EndOfPeriod(d1, flagFeeReceived);
      let i: number;
      for (i = index + 1; i < array.length && array[i].authorId == value.authorId; i++) {
        const d = new Date(array[i].date);
        if (d > d2)
          break;
        value.feeReceived += array[i].feeReceived;
        if (value.id.indexOf('...') < 0) // block fees can be many (from authored blocks)
          value.id += ' ...';
        // value.id += ' ' + array[i].id;
        array[i].ignore = true;
      }
    });
  }

  // 3.3 feePaid entries
  if (flagFeePaid == 'day' || flagFeePaid == 'time') {
    arrFeePaid.forEach((value: TFeePaidRecord, index: number, array: TFeePaidRecord[]) => {
      if (value.ignore) // the entry was processed already
        return;

      const d1 = new Date(value.date);
      const d2 = EndOfPeriod(d1, flagFeePaid);
      let i: number;
      for (i = index + 1; i < array.length && array[i].senderId == value.senderId; i++) {
        const d = new Date(array[i].date);
        if (d > d2)
          break;
        value.feePaid += array[i].feePaid;
        value.id += ' ' + array[i].id;
        array[i].ignore = true;
      }
    });
  }

  // 4. iterate over all records
  console.log('Write csv file...');

  // for progress only:
  const total = arrStaking.length + arrFeeReceived.length + arrFeePaid.length + arrDeposits.length + arrWithdrawals.length;
  const cnt1 = 100.0 * arrStaking.length / total;
  const cnt2 = 100.0 * arrFeeReceived.length / total;
  const cnt3 = 100.0 * arrFeePaid.length / total;
  const cnt4 = 100.0 * arrDeposits.length / total;
  const cnt5 = 100.0 * arrWithdrawals.length / total;

  const staking = OutputStaking(0, cnt1, file, formatStaking, flagStaking == 'day', chainData, arrStaking);
  const feeReceived = OutputFeeReceived(cnt1, cnt2, file, formatFeeReceived, flagFeeReceived == 'day', chainData, arrFeeReceived);
  const feePaid = OutputFeePaid(cnt1 + cnt2, cnt3, file, formatFeePaid, flagFeePaid == 'day', chainData, arrFeePaid);
  const deposits = OutputDeposits(cnt1 + cnt2 + cnt3, cnt4, file, formatDeposit, chainData, arrDeposits);
  const withdrawals = OutputWithdrawals(cnt1 + cnt2 + cnt3 + cnt4, cnt5, file, formatWithdrawal, chainData, arrWithdrawals);

  process.stdout.write('\r  Progress: 100%\n\n');
  console.log(sprintf('Staking records:     %5d   Total: %s %s', staking.count, DivideS(staking.value, chainData.decimals), unit));
  console.log(sprintf('FeeReceived records: %5d   Total: %s %s', feeReceived.count, DivideS(feeReceived.value, chainData.decimals), unit));
  console.log(sprintf('FeePaid records:     %5d   Total: %s %s', feePaid.count, DivideS(-feePaid.value, chainData.decimals), unit));
  console.log(sprintf('Deposit records:     %5d', deposits));
  console.log(sprintf('Withdrawal records:  %5d', withdrawals));

  if (chainData.namesNotFound.length) {
    console.log('\nMissing accounts in accountNames list:');
    chainData.namesNotFound.forEach((account: string) => {
      console.log(account);
    });
  }
}



main();
